File: kotlin/src/commonTest/kotlin/com/actualbudget/sync/crdt/MerkleTest.kt
package com.actualbudget.sync.crdt

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class MerkleTest {

    @Test
    fun testEmptyTrie() {
        val trie = Merkle.emptyTrie()
        assertEquals(0, trie.hash)
        assertTrue(trie.keys().isEmpty())
    }

    @Test
    fun testInsertSingleTimestamp() {
        val ts = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        var trie = Merkle.emptyTrie()
        trie = Merkle.insert(trie, ts)

        assertTrue(trie.hash != 0)
    }

    @Test
    fun testInsertMultipleTimestamps() {
        val ts1 = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val ts2 = Timestamp.parse("2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912")!!

        var trie = Merkle.emptyTrie()
        trie = Merkle.insert(trie, ts1)
        val hash1 = trie.hash

        trie = Merkle.insert(trie, ts2)
        val hash2 = trie.hash

        assertTrue(hash1 != hash2)
    }

    @Test
    fun testBuild() {
        val timestamps = listOf(
            Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!,
            Timestamp.parse("2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912")!!,
            Timestamp.parse("2025-04-24T22:25:42.123Z-0001-A219E7A71CC18912")!!
        )

        val trie = Merkle.build(timestamps)
        assertTrue(trie.hash != 0)
    }

    @Test
    fun testDiffIdenticalTries() {
        val ts = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val trie1 = Merkle.insert(Merkle.emptyTrie(), ts)
        val trie2 = Merkle.insert(Merkle.emptyTrie(), ts)

        assertNull(Merkle.diff(trie1, trie2))
    }

    @Test
    fun testDiffDifferentTries() {
        val ts1 = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val ts2 = Timestamp.parse("2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912")!!

        val trie1 = Merkle.insert(Merkle.emptyTrie(), ts1)
        val trie2 = Merkle.insert(Merkle.emptyTrie(), ts2)

        val diff = Merkle.diff(trie1, trie2)
        assertNotNull(diff)
    }

    @Test
    fun testPrune() {
        val timestamps = (1..100).map { i ->
            Timestamp(1745533422123L + i * 60000, 0, "A219E7A71CC18912")
        }

        var trie = Merkle.emptyTrie()
        for (ts in timestamps) {
            trie = Merkle.insert(trie, ts)
        }

        val pruned = Merkle.prune(trie, 2)
        // Pruned trie should have same hash
        assertEquals(trie.hash, pruned.hash)
    }

    @Test
    fun testSerializeDeserialize() {
        val ts = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val trie = Merkle.insert(Merkle.emptyTrie(), ts)

        val json = Merkle.serialize(trie)
        val restored = Merkle.deserialize(json)

        assertEquals(trie.hash, restored.hash)
    }

    @Test
    fun testKeyToTimestamp() {
        // Base-3 key "0" should give timestamp 0
        assertEquals(0L, Merkle.keyToTimestamp("0"))

        // Non-zero key should give non-zero timestamp
        assertTrue(Merkle.keyToTimestamp("1") > 0)
    }
}

File: dart/test/crdt/merkle_test.dart
import 'package:test/test.dart';
import 'package:actual_sync/src/crdt/merkle.dart';
import 'package:actual_sync/src/crdt/timestamp.dart';

void main() {
  group('Merkle', () {
    test('empty trie', () {
      final trie = Merkle.emptyTrie();
      expect(trie.hash, equals(0));
      expect(trie.keys, isEmpty);
    });

    test('insert single timestamp', () {
      final ts = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      var trie = Merkle.emptyTrie();
      trie = Merkle.insert(trie, ts);

      expect(trie.hash, isNot(equals(0)));
    });

    test('insert multiple timestamps', () {
      final ts1 = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final ts2 = Timestamp.parse('2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912')!;

      var trie = Merkle.emptyTrie();
      trie = Merkle.insert(trie, ts1);
      final hash1 = trie.hash;

      trie = Merkle.insert(trie, ts2);
      final hash2 = trie.hash;

      expect(hash1, isNot(equals(hash2)));
    });

    test('build', () {
      final timestamps = [
        Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!,
        Timestamp.parse('2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912')!,
        Timestamp.parse('2025-04-24T22:25:42.123Z-0001-A219E7A71CC18912')!,
      ];

      final trie = Merkle.build(timestamps);
      expect(trie.hash, isNot(equals(0)));
    });

    test('diff identical tries', () {
      final ts = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final trie1 = Merkle.insert(Merkle.emptyTrie(), ts);
      final trie2 = Merkle.insert(Merkle.emptyTrie(), ts);

      expect(Merkle.diff(trie1, trie2), isNull);
    });

    test('diff different tries', () {
      final ts1 = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final ts2 = Timestamp.parse('2025-04-24T22:24:42.123Z-0001-A219E7A71CC18912')!;

      final trie1 = Merkle.insert(Merkle.emptyTrie(), ts1);
      final trie2 = Merkle.insert(Merkle.emptyTrie(), ts2);

      final diff = Merkle.diff(trie1, trie2);
      expect(diff, isNotNull);
    });

    test('prune', () {
      final timestamps = List.generate(
        100,
        (i) => Timestamp(1745533422123 + i * 60000, 0, 'A219E7A71CC18912'),
      );

      var trie = Merkle.emptyTrie();
      for (final ts in timestamps) {
        trie = Merkle.insert(trie, ts);
      }

      final pruned = Merkle.prune(trie, 2);
      // Pruned trie should have same hash
      expect(pruned.hash, equals(trie.hash));
    });

    test('serialize deserialize', () {
      final ts = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final trie = Merkle.insert(Merkle.emptyTrie(), ts);

      final json = Merkle.serialize(trie);
      final restored = Merkle.deserialize(json);

      expect(restored.hash, equals(trie.hash));
    });

    test('keyToTimestamp', () {
      // Base-3 key "0" should give timestamp 0
      expect(Merkle.keyToTimestamp('0'), equals(0));

      // Non-zero key should give non-zero timestamp
      expect(Merkle.keyToTimestamp('1'), greaterThan(0));
    });
  });
}
